include/libvfio-user.h

```diff
@@ -268,33 +246,106 @@ typedef enum {
+ * type 0 PCI configuration space of the needed size.
+ *
+ * This buffer consists of the standard 64-byte PCI header and the remaining
+ * space depending on @pci_type, either PCI_CFG_SPACE_SIZE or
+ * PCI_CFG_SPACE_EXP_SIZE.
  *
  * @vfu_ctx: the libvfio-user context
- * @id: Device and vendor ID
- * @ss: Subsystem vendor and device ID
- * @cc: Class code
  * @pci_type: PCI type (convention PCI, PCI-X mode 1, PCI-X mode2, PCI-Express)
  * @revision: PCI/PCI-X/PCIe revision
+ */
+vfu_pci_hdr_t *
+vfu_pci_init(vfu_ctx_t *vfu_ctx, vfu_pci_type_t pci_type, int revision UNUSED);
+
+/*
+ * FIXME
+ */
+void
+vfu_pci_set_id(vfu_ctx_t *vfu_ctx, uint16_t vid, uint16_t did,
+               uint16_t ssid, uint16_t ssvid);
+
+/*
+ * FIXME
+ */
+void
+vfu_pci_set_class(vfu_ctx_t *vfu_ctx, uint8_t base, uint8_t sub, uint8_t pi);
+
+// FIXME: what else do we need? multi-function flag in "header type" ? rid?
+
+/**
+ * Add a PCI capability to PCI config space.
+ *
+ * Certain standard capabilities are handled entirely within the library:
+ *
+ * PCI_CAP_ID_EXP (pxcap)
+ * PCI_CAP_ID_MSIX (msixcap)
+ * PCI_CAP_ID_PM (pmcap)
+ *
+ * However, they must still be explicitly initialized and added here.
  *
- * @returns 0 on success, -1 on failure and sets errno.
+ * FIXME: are we really making the user init these too? why? Is there a better
+ * API?
+ *
+ * The contents of @data is copied in. It must start with either a struct
+ * pci_cap or a struct pci_extcap, with the ID field set (the 'next' field is
+ * ignored).  If the capability is not one of the handled ones above (e.g.
+ * PCI_CAP_ID_VNDR), the library will allow all reads and disallow all writes
+ * for the capability from the client.
+ *
+ * Note the size of the capability is derived either from the ID, or from an
+ * embedded length field for PCI_(EXT_)CAP_ID_VNDR.
+ *
+ * @vfu_ctx: the libvfio-user context
+ * @cap: capability ID to add
+ * @extended: true if an extended capability (PCI_EXT_CAP_ID_*)
+ * @data: capability data, not including the capability header
  */
 int
-vfu_pci_setup_config_hdr(vfu_ctx_t *vfu_ctx, vfu_pci_hdr_id_t id,
-                         vfu_pci_hdr_ss_t ss, vfu_pci_hdr_cc_t cc,
-                         vfu_pci_type_t pci_type,
-                         int revision __attribute__((unused)));
+vfu_pci_add_capability(vfu_ctx_t *vfu_ctx, uint16_t cap_id, bool extended,
+                       void *data);
+
+/**
+ * Callback function that gets called when a capability is accessed.
+ *
+ * @pvt: private pointer
+ * @id: capability ID being accessed
+ * @buf: pointer to data being read or written
+ * @count: number of bytes being read or written
+ * @offset: offset within the capability (including the header)
+ * @is_write: whether the capability is read or written
+ *
+ * @returns the number of bytes read or written
+ *
+ * FIXME: cb would have to identify each cap by "pvt" - is that OK?
+ * FIXME: this can already be handled by region callback - do we need this
+ * too?
```

Comment:
The region callback (`vfu_region_access_cb_t`) doesn't have the capability ID, so we need `vfu_cap_access_cb_t`. Alternatively, we can add a region parameter to `vfu_region_access_cb_t` so when a region is accessed this parameter will be the region and when a capability is accessed it will be tha capability ID.

Reply:
Well, there's nothing stopping a user parsing the cap list to figure this out, that's what I meant. It's somewhat weird to have a cb for caps and the config space region when one is container in the other.

Reply:
I see what you mean. I think we shouldn't have a callbacl for the PCI config space region at all, everything should be handled by the library. Then having callbacks specific for PCI capabilites (assuming we'll need them) is necessary.

Reply:
This comes down to "need more understanding of typical devices", but there's free space in config space that could be completely arbitrary device area, isn't there?

Reply:
I wasn't clear enough, I meant that there shouldn't be a callaback for the PCI config space _header_. I don't know whether the remaining space must be used exclusively for capabilities or the device and device driver can use this space as they please. I wouldn't think that would be the case though, if they have to do that they would use the vendor-specific capability. To make further progress on this we need to undrstand more about PCI, I've started compiling a list of questions here: https://github.com/nutanix/libvfio-user/issues/180.

Reply:
My thinking was generally that we'd handle what we knew how to, and punt the rest to the config space cb if there is one.